src/nacl/bindings/crypto_secretstream.py

```diff
@@ -0,0 +1,271 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import, division, print_function
+
+from nacl import exceptions as exc
+from nacl._sodium import ffi, lib
+from nacl.exceptions import ensure
+
+
+crypto_secretstream_xchacha20poly1305_ABYTES = \
+    lib.crypto_secretstream_xchacha20poly1305_abytes()
+crypto_secretstream_xchacha20poly1305_HEADERBYTES = \
+    lib.crypto_secretstream_xchacha20poly1305_headerbytes()
+crypto_secretstream_xchacha20poly1305_KEYBYTES = \
+    lib.crypto_secretstream_xchacha20poly1305_keybytes()
+crypto_secretstream_xchacha20poly1305_MESSAGEBYTES_MAX = \
+    lib.crypto_secretstream_xchacha20poly1305_messagebytes_max()
+crypto_secretstream_xchacha20poly1305_STATEBYTES = \
+    lib.crypto_secretstream_xchacha20poly1305_statebytes()
+
+
+crypto_secretstream_xchacha20poly1305_TAG_MESSAGE = \
+    lib.crypto_secretstream_xchacha20poly1305_tag_message()
+crypto_secretstream_xchacha20poly1305_TAG_PUSH = \
+    lib.crypto_secretstream_xchacha20poly1305_tag_push()
+crypto_secretstream_xchacha20poly1305_TAG_REKEY = \
+    lib.crypto_secretstream_xchacha20poly1305_tag_rekey()
+crypto_secretstream_xchacha20poly1305_TAG_FINAL = \
+    lib.crypto_secretstream_xchacha20poly1305_tag_final()
+
+
+def crypto_secretstream_xchacha20poly1305_keygen():
+    """
+    Generate a key for use with
+    :func:`.crypto_secretstream_xchacha20poly1305_init_push`.
+
+    """
+    keybuf = ffi.new(
+        "unsigned char[]",
+        crypto_secretstream_xchacha20poly1305_KEYBYTES,
+    )
+    lib.crypto_secretstream_xchacha20poly1305_keygen(keybuf)
+    return ffi.buffer(keybuf)[:]
+
+
+def crypto_secretstream_xchacha20poly1305_init_push(key):
+    """
+    Initialize a crypto_secretstream_xchacha20poly1305 encryption buffer.
+
+    :param key: must be
+                :data:`.crypto_secretstream_xchacha20poly1305_KEYBYTES` long
+    :type key: bytes
+    :return: (statebuf, header)
+    :rtype: (bytes, bytes)
+    """
+    ensure(
+        isinstance(key, bytes),
+        'Key must be a bytes sequence',
+        raising=exc.TypeError,
+    )
+    ensure(
+        len(key) == crypto_secretstream_xchacha20poly1305_KEYBYTES,
+        'Invalid key length',
+        raising=exc.ValueError,
+    )
+
+    statebuf = ffi.new(
+        "unsigned char[]",
+        crypto_secretstream_xchacha20poly1305_STATEBYTES,
+    )
+
+    # both _salt and _personal must be zero-padded to the correct length
+    headerbuf = ffi.new(
+        "unsigned char []",
+        crypto_secretstream_xchacha20poly1305_HEADERBYTES,
+    )
+
+    rc = lib.crypto_secretstream_xchacha20poly1305_init_push(
+        statebuf, headerbuf, key)
+    ensure(rc == 0, 'Unexpected failure', raising=exc.RuntimeError)
+
+    return statebuf, ffi.buffer(headerbuf)[:]
```

Comment:
I cribbed this pattern from `generichash_blake2b_init`. I'd probably err toward just dropping the type documentation instead of actually fixing this. I agree it's a little bit weird but it's an opaque pointer so whatever you want to do about it is fine by me. Let me know what you think.

Reply:
Ah, that was exactly what I was looking for, I missed that was doing the same thing there. That seems fine to me then since it's consistent with what's already being used in the library.

Reply:
The alternative I considered was something similar to `crypto_sign_ed25519ph_state`. Do you prefer that? The only benefit I see is that it lets me attach some docs to the returned object but I'm meh on it.

Reply:
Nah I think this is OK.

Reply:
I think we should correct the return type (or at least not say it's bytes) for both this and generichash_init, but we can do that in a followup PR.

Reply:
I pushed a new version with some tweaks to indicate `object`. A separate commit also fixes the blake2b module.

Reply:
I'm going to update this PR with a state object. The reason I went against it was that there was prior-art in pynacl and so I copied that, but I have come to realize that there are some optimizations I can make if I go with the object approach such as caching a byte buffer to re-use instead of allocating new memory per push/pull. Stay tuned!